plans/data-transfer-seeds-leeches/utils/createnode.go

```diff
@@ -0,0 +1,155 @@
+}
+
+func (n *Node) Close() {
+	n.Bitswap.Close()
+}
+
+// CreateNode creates a libp2p Node with a Bitswap instance
+func CreateNode(ctx context.Context, runenv *runtime.RunEnv) (*Node, error) {
+	bstoreDelay := runenv.IntParamD("bstore_delay_ms", 0)
+
+	h, err := libp2p.New(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	routing, err := nilrouting.ConstructNilRouting(nil, nil, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	net := bsnet.NewFromIpfsHost(h, routing)
+
+	bsdelay := delay.Fixed(time.Duration(bstoreDelay) * time.Millisecond)
+	dstore := ds_sync.MutexWrap(delayed.New(ds.NewMapDatastore(), bsdelay))
+	bstore, err := blockstore.CachedBlockstore(ctx,
+		blockstore.NewBlockstore(ds_sync.MutexWrap(dstore)),
+		blockstore.DefaultCacheOpts())
+	if err != nil {
+		return nil, err
+	}
+
+	bitswap := bs.New(ctx, net, bstore).(*bs.Bitswap)
+	bserv := blockservice.New(bstore, bitswap)
+	dserv := merkledag.NewDAGService(bserv)
+	return &Node{h, bitswap, dserv}, nil
+}
+
+type AddSettings struct {
+	Layout    string
+	Chunker   string
+	RawLeaves bool
+	Hidden    bool
+	NoCopy    bool
+	HashFunc  string
+	MaxLinks  int
+}
+
+func (n *Node) Add(ctx context.Context, r io.Reader) (ipld.Node, error) {
+	settings := AddSettings{
+		Layout:    "balanced",
+		Chunker:   "size-262144",
+		RawLeaves: false,
+		Hidden:    false,
+		NoCopy:    false,
+		HashFunc:  "sha2-256",
+		MaxLinks:  helpers.DefaultLinksPerBlock,
+	}
+	// for _, opt := range opts {
+	// 	err := opt(&settings)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// }
+
+	prefix, err := merkledag.PrefixForCidVersion(1)
+	if err != nil {
+		return nil, errors.Wrap(err, "unrecognized CID version")
+	}
+
+	hashFuncCode, ok := multihash.Names[strings.ToLower(settings.HashFunc)]
+	if !ok {
+		return nil, errors.Wrapf(err, "unrecognized hash function %q", settings.HashFunc)
+	}
+	prefix.MhType = hashFuncCode
+
+	dbp := helpers.DagBuilderParams{
+		Dagserv:    n.Dserv,
+		RawLeaves:  settings.RawLeaves,
+		Maxlinks:   settings.MaxLinks,
+		NoCopy:     settings.NoCopy,
+		CidBuilder: &prefix,
+	}
+
+	chnk, err := chunker.FromString(r, settings.Chunker)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create chunker")
+	}
+
+	dbh, err := dbp.New(chnk)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create dag builder")
+	}
+
+	var nd ipld.Node
+	switch settings.Layout {
+	case "trickle":
+		nd, err = trickle.Layout(dbh)
+	case "balanced":
+		nd, err = balanced.Layout(dbh)
+	default:
+		return nil, errors.Errorf("unrecognized layout %q", settings.Layout)
+	}
+
+	return nd, err
+}
+
+func (n *Node) FetchGraph(ctx context.Context, c cid.Cid) error {
+	ng := merkledag.NewSession(ctx, n.Dserv)
+	return Walk(ctx, c, ng)
+}
+
+func (n *Node) Get(ctx context.Context, c cid.Cid) (files.Node, error) {
+	nd, err := n.Dserv.Get(ctx, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get file %q", c)
+	}
+
+	return unixfile.NewUnixfsFile(ctx, n.Dserv, nd)
+}
```

Comment:
Woa! This is super pretty advanced go-ipfs node creation!

Don't let me distract you from what you need to test in the first place, but would using [go-ipfs as a library](https://github.com/ipfs/go-ipfs/tree/master/docs/examples/go-ipfs-as-a-library) make it simpler for you and for future maintainers?

You can check my create-node-util from https://github.com/ipfs/testground/pull/58/files#diff-2965c8a82c065af5207a99292783825d

Reply:
That's a good question - I wanted to isolate Bitswap as much as possible for this particular test, so it's easier to track down any problems. If we use go-ipfs as a library it brings in a lot of other code so.

Reply:
I should add that I think it makes sense to use go-ipfs as a library for the scenarios that are good for canary testing like https://github.com/ipfs/testground/tree/master/plans/data-transfer-datasets-random

Reply:
One other thing to mention is that I ported over the custom DAG walking function from p2plab. This allows us to maximize the number of concurrent block requests to Bitswap (basically just request every linked CID in a block as soon as we receive the block). Depending on how IPFS fetch is invoked (get / pin / etc) it has request patterns that are more conservative, only pre-fetching a select group of linked blocks at a time. By maximizing the concurrent block requests we can distinguish between the performance of Bitswap vs Bitswap within IPFS.